do not use smart pointer references

ptschip opened the pull request:
Not sure if I'm being a bit pedantic here but I think it's a bad practice to be using references to smart pointers in a multi-threaded environment, otherwise why bother using smart pointers? I think we'll just end up getting bitten by it down the road when someone thinks they've got a valid object but it's already gone out of scope in some other thread..

ptschip committed:
Do not pass smart pointers as a reference in policy.cpp/h

ptschip committed:
Do not pass smart pointers as a reference in txmempool.cpp/h

ptschip committed:
Do not pass smart pointers as a reference in respend .cpp/h files

ptschip committed:
Do not pass smart pointers as a reference in thinblock.cpp/h

gandrewstone commented:
I think that this is likely a bit slower than passing by reference but we should do it and optimize (probably by passing by pointer, actually) specific functions only based on real performance and profiling test results.

ptschip commented:
very odd that this seems to break a lot of py tests.. have to investigate this.

ptschip commented:
ok, found the problem...forgot about when we pass by reference to get a result back.

ptschip committed:
Do not pass smart pointers as a reference in tx_verify.cpp/h

ptschip committed:
Do not pass smart pointers as a reference in txorphanpool.cpp/h

ptschip committed:
Do not pass smart pointers as a reference in net.cpp/h

ptschip committed:
Do not pass smart pointers as a reference in txadmission.cpp/h

ptschip committed:
Fix compactblocks_2.py

Add a check for just checking the mempool() rather than both
mempool and orphanpool.
